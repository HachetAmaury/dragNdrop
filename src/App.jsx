import React from 'react';

import './AppStyle.less';

const SQUARES_PER_LINE = 4;

class App extends React.Component {
    constructor() {
        super();

        // The squares are just a list of rgb colors strings ( ex : "rgb(82,232,76)")
        // Since the only important property for a square is its color
        const squares = this.createAllSquares();
        this.state = { squares };
    }

    getTotalSquares = () => {
        return SQUARES_PER_LINE * SQUARES_PER_LINE;
    };

    getRandomColorCode() {
        return ~~(Math.random() * 256);
    }

    getOneRandomRGBColor = () => {
        return `rgb(${this.getRandomColorCode()},${this.getRandomColorCode()},${this.getRandomColorCode()})`;
    };

    createAllSquares = () => {
        const squares = [];
        const squaresAmount = this.getTotalSquares();

        for (let index = 0; index < squaresAmount; index++) {
            squares.push(this.getOneRandomRGBColor());
        }

        return squares;
    };

    onDragStart = (ev, index) => {
        // Store the square index in the squares array
        // for later use in the onDrop callback
        ev.dataTransfer.setData('index', index);
    };

    onDragOver = ev => {
        // To allow the onDrop event to be fired
        ev.preventDefault();
    };

    onDrop = (ev, newIndex) => {
        // Retrieve the index of the dropped square
        const oldIndex = ev.dataTransfer.getData('index');
        this.switchSquares(oldIndex, newIndex);
        this.setState(this.state);
    };

    switchSquares = (oldIndex, newIndex) => {
        const tempData = this.state.squares[oldIndex];
        this.state.squares[oldIndex] = this.state.squares[newIndex];
        this.state.squares[newIndex] = tempData;
    };

    getOneSquareColor = squareIndex => {
        return this.state.squares[squareIndex];
    };

    displaySquares = () => {
        const squares = [];
        const totalSquares = this.getTotalSquares();

        for (let index = 0; index < totalSquares; index++) {
            const squareColor = this.getOneSquareColor(index);
            squares.push(
                <div
                    key={index}
                    id={`square-${index}`}
                    draggable
                    onDragStart={e => {
                        this.onDragStart(e, index);
                    }}
                    onDragOver={e => this.onDragOver(e)}
                    onDrop={e => {
                        this.onDrop(e, index);
                    }}
                    className="square"
                    style={{ background: squareColor }}
                ></div>,
            );
        }

        return squares;
    };

    render() {
        return (
            <div className="app">
                <div className="container">
                    <div className="title-container">
                        <div className="title"> Drag & Drop</div>
                    </div>
                    <div className="board-container">
                        <div
                            className="board"
                            style={{
                                gridTemplateColumns: `repeat(${SQUARES_PER_LINE}, 1fr)`,
                                gridTemplateRows: `repeat(${SQUARES_PER_LINE}, 1fr)`,
                            }}
                        >
                            {this.displaySquares()}
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

export default App;
