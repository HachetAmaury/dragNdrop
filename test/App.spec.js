import React from 'react';
import { shallow, mount } from 'enzyme';
import App from '../src/App';

describe('Application', () => {
    it('should render without throwing an error', () => {
        expect(shallow(<App />)).toMatchSnapshot();
    });

    it('should display the correct amount of squares', () => {
        const wrapper = mount(<App />);

        expect(wrapper.find('div.app')).toHaveLength(1);
        expect(wrapper.find('div.square')).toHaveLength(16);

        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });

    it('should getOneRandomRGBColor correctly ', () => {
        const wrapper = mount(<App />);

        expect(wrapper.instance().getOneRandomRGBColor()).toMatch(
            /rgb(.*,.*,.*)/,
        );

        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });

    it('should createAllSquares correctly ', () => {
        const wrapper = mount(<App />);

        expect(wrapper.instance().createAllSquares()).toHaveLength(16);

        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });

    it('should switch first and second squares', () => {
        const wrapper = mount(<App />);

        const firstSquareBackground = wrapper.find('#square-0').prop('style')
            .background;
        const secondSquareBackground = wrapper.find('#square-1').prop('style')
            .background;

        wrapper.instance().switchSquares(0, 1);

        wrapper.setProps({});

        expect(wrapper.find('#square-0').prop('style').background).toEqual(
            secondSquareBackground,
        );

        expect(wrapper.find('#square-1').prop('style').background).toEqual(
            firstSquareBackground,
        );

        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });

    it('should switch first and second then second and first squares and be unchanged', () => {
        const wrapper = mount(<App />);

        const firstSquareBackground = wrapper.find('#square-0').prop('style')
            .background;
        const secondSquareBackground = wrapper.find('#square-1').prop('style')
            .background;

        wrapper.instance().switchSquares(0, 1);
        wrapper.instance().switchSquares(1, 0);

        wrapper.setProps({});

        expect(wrapper.find('#square-0').prop('style').background).toEqual(
            firstSquareBackground,
        );

        expect(wrapper.find('#square-1').prop('style').background).toEqual(
            secondSquareBackground,
        );

        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });
});
