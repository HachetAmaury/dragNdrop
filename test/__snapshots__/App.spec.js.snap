// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application should createAllSquares correctly  1`] = `ReactWrapper {}`;

exports[`Application should display the correct amount of squares 1`] = `ReactWrapper {}`;

exports[`Application should getOneRandomRGBColor correctly  1`] = `ReactWrapper {}`;

exports[`Application should render without throwing an error 1`] = `ShallowWrapper {}`;

exports[`Application should render without throwing an error 2`] = `ReactWrapper {}`;

exports[`Application should switch first and second squares 1`] = `ReactWrapper {}`;

exports[`Application should switch first and second then second and first squares and be unchanged 1`] = `ReactWrapper {}`;
